spark3/low-cpu-usage.not-enough-partitions-in-input.scala

```diff
@@ -0,0 +1,122 @@
+
+*/
+
+// COMMAND ----------
+
+
+import java.util.UUID
+import org.apache.spark.sql.{DataFrame, SparkSession}
+
+val spark: SparkSession = SparkSession.active
+
+spark.conf.set("spark.sql.adaptive.enabled", false)
+val input = "/tmp/perf-hikes/datasets/optd_por_public_filtered.csv"
+val tmpPath = "/tmp/perf-hikes/sandbox/" + UUID.randomUUID()
+
+spark.sparkContext.setJobDescription("Read CSV")
+val airports = spark.read.option("delimiter","^").option("header","true").csv(input)
+
+def showParitions(df: DataFrame) = df.withColumn("part_id", spark_partition_id()).groupBy("part_id").count().selectExpr("count(distinct(part_id)) as legit_partitions", "max(count) as max_records", "min(count) as min_records").show()
+
+// COMMAND ----------
+
+spark.sparkContext.setJobDescription("Initialize input table")
+spark.conf.set("parquet.block.size", 1024 * 1024) // size of the row-group is 1MB
+
+// single parquet file, roughly 4MB each (we create multiple input datasets to avoid warmup effects)
+Range(1,10).map(r => airports.withColumn("idx", lit(r))).reduce(_.union(_)).repartition(1).write.format("parquet").save(tmpPath + "/input1")
+Range(1,10).map(r => airports.withColumn("idx", lit(r))).reduce(_.union(_)).repartition(1).write.format("parquet").save(tmpPath + "/input2")
+Range(1,10).map(r => airports.withColumn("idx", lit(r))).reduce(_.union(_)).repartition(1).write.format("parquet").save(tmpPath + "/input3")
+
+// COMMAND ----------
+
+val expensiveProcessing = "sha(sha(sha(sha(sha(sha(sha(sha(sha(sha(sha(sha(sha(sha(sha(sha(sha(sha(sha(sha(sha(sha(wiki_link))))))))))))))))))))))"
+
+// COMMAND ----------
+
+// Scenario with few partitions
+spark.sparkContext.setJobDescription("Read input (few partitions)")
+val df1 = spark.read.format("parquet").load(tmpPath + "/input1")
+showParitions(df1) // what's the distribution of records per partition?
+
+spark.sparkContext.setJobDescription("Write input (few partitions)")
+df1.selectExpr(expensiveProcessing).write.format("noop").mode("overwrite").save()
+
+// COMMAND ----------
+
+// Scenario with many partitions
+spark.conf.set("spark.sql.files.maxPartitionBytes", 128*1024)
+spark.sparkContext.setJobDescription("Read input (many partitions)")
+val df2 = spark.read.format("parquet").load(tmpPath + "/input2")
+showParitions(df2) // what's the distribution of records per partition?
+
+spark.sparkContext.setJobDescription("Write input (many partitions)")
+df2.selectExpr(expensiveProcessing).write.format("noop").mode("overwrite").save()
+// It is possible to see the setting in the Spark UI, Tab 'SQL / DataFrame' then 'SQL / DataFrame Properties'
+
+// COMMAND ----------
+
+// Scenario with repartition
+spark.sparkContext.setJobDescription("Read input (repartitioned)")
+val df3 = spark.read.format("parquet").load(tmpPath + "/input3")
+spark.sparkContext.setJobDescription("Write input (repartitioned)")
+df3.repartition(4).selectExpr(expensiveProcessing).write.format("noop").mode("overwrite").save()
```

Comment:
Wondering if we should elaborate on advice on strategies how to pick this number?
- fixing a target max size per partition and obtaining the input size from the optimized plan stats
- increasing to a number which is a multiple of the available cores